lib/offline_sort/offline_sort.rb

```diff
@@ -11,42 +11,57 @@ class Sorter
     DEFAULT_CHUNK_IO_CLASS = defined?(::MessagePack) ? Chunk::InputOutput::MessagePack : Chunk::InputOutput::Marshal
     DEFAULT_CHUNK_SIZE = 1000
 
-    attr_reader :enumerable, :sort_by, :chunk_size, :chunk_input_output_class
+    attr_reader :enumerable, :sort_by, :chunk_size, :chunk_input_output_class, :sorted_chunks
 
     def initialize(enumerable, chunk_input_output_class: DEFAULT_CHUNK_IO_CLASS, chunk_size: DEFAULT_CHUNK_SIZE, &sort_by)
       @enumerable = enumerable
       @chunk_input_output_class = chunk_input_output_class
       @chunk_size = chunk_size
       @sort_by = sort_by
+      @sorted_chunks = []
     end
 
-    def sort
-      sorted_chunks = []
-      chunk_entries = []
+    # Sorts the enumerable passed, maintaining a max memory of chunk size
+    #
+    # @param close_tempfiles [Boolean]
+    #   Whether or not to close the tempfiles opened by the sort. Note if this
+    #   method will be called multiple times for the same Sorter object, every
+    #   time except for the last must pass close_tempfiles as false.
+    #
+    # @return [Enumerator] The sorted enumerable
+    def sort(close_tempfiles: true)
+      if sorted_chunks.empty?
```

Comment:
With the branching that's happening here, it feels a little bit like we've got two different functions combined into one here. I'm wondering if it may help to introduce a new `Merger` class and have the behavior be something like:

```
def sort
   < all the logic to create sorted chunks >

   Merger.new(sorted_chunks)
end
```

```
class Merger
   def initialize(sorted_chunks)
      @sorted_chunks = sorted_chunks
   end

   delegate :each, to: :enumearator

    def enumerator(sorted_chunk_ios)
      pq = []
      chunk_enumerators = sorted_chunk_ios.each(&:open).each(&:rewind).map(&:each)

      chunk_enumerators.each_with_index do |chunk, index|
        entry = chunk.next
        pq.push(ChunkEntry.new(index, entry))
      end

      entry_sort_by = Proc.new { |entry| sort_by.call(entry.data) }
      pq = FixedSizeMinHeap.new(pq, &entry_sort_by)

      Enumerator.new do |yielder|
        while item = pq.pop
          yielder.yield(item.data)

          begin
            entry = chunk_enumerators[item.chunk_number].next
            pq.push(ChunkEntry.new(item.chunk_number, entry))
          rescue StopIteration
            sorted_chunk_ios[item.chunk_number].close
          end
        end
      end
end
```

That way in the code if you do:

```
x = OfflineSort.sort(list)
```

`x` will either be an in memory array which you can call `each` on multiple times or will be a `Merger` which you can also call `each` on multiple times.

Note: instead of passing a `close_tempfiles` arg, the code here just closes always and then reopens/rewinds at the beginning of the block:

```
irb(main):012:0> x = Tempfile.new
=> #<Tempfile:/var/folders/_4/2jtv2cpj3tjc66m3vlj1k_y40000gn/T/20200902-19202-n3za3>
irb(main):013:0> x.write('hello')
=> 5
irb(main):014:0> x.close
=> nil
irb(main):015:0> x.open
=> #<File:/var/folders/_4/2jtv2cpj3tjc66m3vlj1k_y40000gn/T/20200902-19202-n3za3>
irb(main):016:0> x.rewind
=> 0
irb(main):017:0> x.read
=> "hello"
irb(main):018:0>
```